### ISSUE TITLE
Consolidate certification properties in database? same with profile properties?

### ISSUE THREAD

**Issue by moT01** [freeCodeCamp/freeCodeCamp#48906]
Creating this issue based on a comment from another issue...
<details>
<summary>original comment</summary>
        Another thought I have is that it seems kind of bloated to include all 15 or so of the certificates and properties on the top level of the user. Besides cluttering out the other properties instead of being encapsulated into one property with sub properties, it makes our code more clunky as well. Consider the current reset logic, most of the lines are just dedicated to resetting the certificates and if any new certificates are added it would be easy to forget them here, which would create bugs. 

![Screen Shot 2022-12-24 at 6 37 47 PM](https://user-images.githubusercontent.com/20331238/209454751-a7ca0b21-5dce-4f46-ba73-28f354574056.png)

Wouldn't it be cleaner to put them within an object property on the user? I see that the `profileUI` property on the user is similar. That would be the ideal in my opinion. If we did that then in this reset function we could just set the value to an empty object which would be much more maintainable.

Another way that this makes things more difficult is the strange naming and the mapping required to work with it. I'm talking about things like `api-server/src/server/utils/certTypes.json`. For this task for this task (if we don't do this proposed improvement to simplify things) I might have to create an additional mapping on top of that sense even the keys of that mapping don't match the IDs of the superblocks.

However, If we make the change I'm proposing and had an object property on the user the keys of that object would be the IDs of the super blocks. This would eliminate the need for mappings like the one above.

_Originally posted by @oriooctopus in https://github.com/freeCodeCamp/freeCodeCamp/issues/47464#issuecomment-1364609695_
</details>

The gist of the suggestion is that we should consolidate all the "isCert" (e.g. `isFrontEndCert`, etc) into an object or array on the user object in the database - and maybe do the same with the profile properties.

More related comments on that same issue:
`https://github.com/freeCodeCamp/freeCodeCamp/issues/47464#issuecomment-1368567247`
`https://github.com/freeCodeCamp/freeCodeCamp/issues/47464#issuecomment-1368567949`

**Comment by ojeytonwilliams** [comment]
I think we generally agree that the current approach is confusing to work with.  However, as ever with the db, we need to figure out a way of migrating, ideally without downtime.

In principle this shouldn't be too hard, since the existing records all either have `isXCert` or they don't and it's reasonably safe to assume that their certification status is not stored some other way. So, if we expand the api to handle both structures, run a migration script and then contract the api to only handle the new approach that should work.

That said, @naomi-lgbt and @raisedadead have a better handle on Mongodb (and our db in particular) than I do, so they may have thoughts on the feasibility of something like this.

**Comment by raisedadead** [comment]
Has been in discussion for ages, only no one liked the idea when I raised it years ago https://github.com/freeCodeCamp/freeCodeCamp/issues/38609

**Comment by ojeytonwilliams** [comment]
I think this is more in line with another discussion we had, @raisedadead. I'm not sure if there's an issue documenting it, but I feel like we've talked about something like

```js
{
  _id: ObjectId('5bd30e0f1caf6ac3ddddddb5'),
  email: 'foo@bar.com',
  ...
  certifications: {
    isFrontEndCert: false,
    ...
  }
}
```
before.

We can talk more about the precise structure (potentially including the ideas in the linked issue), but I figured we'd want to sketch out a migration plan first.  After all, if we're not confident we can migrate, deciding which schema we want can wait until we are.

**Comment by raisedadead** [comment]
Someone bring it up in the next dev meeting.

**Comment by oriooctopus** [comment]
Was there any discussion around this at the dev meeting?

**Comment by moT01** [comment]
Yes @oriooctopus, it sounds like this is something we want. The steps we discussed to accomplish this are:

1. Migrate the database - add the new data and keep the old. So basically, create the new data we want while keeping the existing data
2. Add/change the API code to read the new data
3. Start using the new data
4. Delete the original data in the database

I'm not sure if we want to try and tackle both the certification data and the profile data at the same time or separate them and do them one at a time. I feel like I would separate them to make it easier.

**Comment by oriooctopus** [comment]
@moT01 I won't have time to do this unfortunately

**Comment by raisedadead** [comment]
Yeah - this is one of those things that are most likely to get developed by the staff-dev team. No worries we appreciate you thoughts and insights in many of the issues and PRs :)

Happy coding.

**Comment by moT01** [comment]
Closing as stale.
